[spark-21933][yarn] Spark Streaming request more executors than excepted without DynamicAllocation

klion26 opened the pull request:
## What changes were proposed in this pull request?

Add logic to test whether the complete container has been reported complete before

## How was this patch tested?

manual tests 

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

Please review http://spark.apache.org/contributing.html before opening a pull request.

Commit:
test whether the complete container has been completed before

AmplabJenkins commented:
Can one of the admins verify this patch?

klion26 commented:
create a issue https://issues.apache.org/jira/browse/SPARK-21933

HyukjinKwon commented:
Could you fix the title to be a form, `[SPARK-xxxx][COMPONENT] Title`, as described in http://spark.apache.org/contributing.html?

klion26 commented:
@HyukjinKwon i am sorry for that, have changed the title form

klion26 commented:
@HyukjinKwon  @vanzin @srowen @foxish @djvulee @squito Could you please help to review this pr?

klion26 commented:
Will the same completed message will be reported more than twice, if these message will not be reported more than twice, then i could use `completedContainerIdSet.remove(containerId)` 
 instead of `completedContainerIdSet.contains(containerId)` to save memory.

Could any one can help to review this patch?

klion26 commented:
@jerryshao Could you help me to review this pathc?

caneGuy commented on resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -158,6 +158,7 @@ private[yarn] class YarnAllocator(
   private[yarn] val containerPlacementStrategy =
     new LocalityPreferredContainerPlacementStrategy(sparkConf, conf, resource, resolver)
 
+  private[yarn] val completedContainerIdSet: HashSet[ContainerId] = new HashSet[ContainerId]()
```
Any recycling strategy？I wonder there may be subtle issue of memory consuming.

klion26 replied:
My friend told me the completedContainer will only be reported twice, if that's true, we could use `completedContainerIdSet.remove(containerId)` instead of `completedContainerIdSet.contains(containerId)` as I said above.

But I can't confirm that, I don't know much about Yarn.

jerryshao commented:
Hi @klion26 , sorry for the late response. Can we please understand the problem first, would you please describe your problem in detail and how to reproduce your issue?

klion26 commented:
Hi @jerryshao, thank you for your reply.

# Problem
the problem is for long running jobs which run on **yarn with HA** will request more executors than it requests.

# How to reproduce 
1. start a spark streaming job on yarn
2. mark one of the nodemanagers which runs container of the spark streaming program as lost(this step will take 10 minutes in my environment)
3. the nodemanger which lost in step 2 came back
4. restart the resourcemanager
5. after the resourcemanger restarted, we will get more resource than we request


# Question
I have one question: should i use `completedContainerIdSet.remove(containerId)` instead of `completedContainerIdSet. contains(containerId)`, if the container lost message will only be reported twice, we should use `remove` instead of `contains` method

jerryshao commented:
Did you enable RM or NM recovery, can you please clarify it?

Normally, if we assume there's are 2 containers running on this NM, after 10 minutes, RM will detect the failure of NM and relaunch 2 lost containers in other NMs, and the total number of executors should still be the same. But things will be different if we enabled NM recovery, because now the failure of NM will not lead to container lost.

klion26 commented:
We enabled RM and NM recovery.

If we assume there are 2 containers running on this NM, after 10 minute, RM detects the failure of NM and relaunches 2 lost containers in other NMs. This is ok. 

But if we restart the RM, then, the lost containers in the NM will be **reported to RM as lost again** because of recovery, we will relaunch 2 more containers in other NMs, then we will get 2 more executors than we expected.

jerryshao commented:
>But if we restart the RM, then, the lost containers in the NM will be reported to RM as lost again because of recovery

Since you already enabled RM and NM recovery, IIUC the failure of RM/NM will not lead to container exit. And after RM/NM restart, it will recover the persistent container metadata, so I think there should be no lost containers reported. Sorry I'm not so familiar with this part in YARN.

jerryshao commented:
And based on your fix:

1. looks like you don't have retention mechanism, which will potential introduce memory leak.
2. I don't see your logic to avoid requesting new containers, is your current logic enough to fix the issue.

Commit:
use HashSet.remove() instead of HashSet.contains() to memory reclaim

klion26 commented:
My colleague create a [issue](https://issues.apache.org/jira/browse/YARN-7214) here, I rewrite the description here.

Spark Streaming (app1) running on Yarn, app1's one container (c1) runs on NM1.
1. NM1 crashed, and RM found NM1 expired in 10 minutes.
2. RM will remove all containers in NM1(RMNodeImpl). and app1 will receive completed message of c1. But RM can not send c1(to be removed) to NM1 because NM1 lost.
3. NM1 restart and register with RM(c1 in register request), but RM found NM1 is lost and will not handle containers from NM1.
4 NM1 will not heartbeat c1(c1 not in heartbeat request). So c1 will not removed from context of NM1.
5. RM restart, NM1 re register with RM. And c1 will handled and recovered. RM will send c1 completed message to AM of app1. So, app1 received a duplicated completed message of c1.


For the fix
1. I changed the code from `completedContainerIdSet.contains(containerId)` to `completedContainerIdSet.remove(containerId)` to reclaim the memory. (The same container will not reported as completed more than twice)
2. The code I added is to ignore the duplicated completed messages, ignore the completed message will avoid requesting new containers.

squito commented:
I'm not sure I totally follow the sequence of events, but I get the feeling this should be handled in yarn, not spark.

Also, I agree with Jerry, it seems like your `completedContainerIdSet` may grow continuously.  You'll remove from it *if* you happen to get a duplicate message.  But I think in most cases you will not a get duplicate message, if I understand correctly.

klion26 commented:
@squito I agree with you that this should be handled by yarn. 

In my opinion, this is some form of defensive programming. The Spark Streaming and structured streaming will both request more resource than they want, if these things happen.

klion26 commented:
@jerryshao thank you for your comment, I will try to find how MR/TEZ handle this

vanzin commented:
Have you guys reached a consensus on whether this PR is needed or not?

klion26 commented:
sorry for the late response, IIUC, in MR, this case handled by below
1. AM received the container failed message
2. AM will check whether there are any attempts of the same task is RUNNING or SUCCEED
2.1 If step 2 returns true, then MR ignores the failed message
2.2 if step 2 returns false, then MR will request a new container and run the specified task

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

vanzin commented:
Looks like the consensus here is that this should not be dealt in Spark. Given this doesn't merge and has unaddressed feedback regarding increasing memory usage, I'm going to close this for now.

If you update your branch, then this will automatically reopen.
